File: app/src/main/java/com/boardgamegeek/ui/widget/ScoreGraphView.java
package com.boardgamegeek.ui.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.support.annotation.ColorInt;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;

import com.boardgamegeek.R;

public class ScoreGraphView extends View {
	private Paint barPaint;
	private Paint scorePaint;
	private float scoreRadius;
	private float smallTickHeight;
	private float largeTickHeight;
	private int lowScoreColor;
	private int averageScoreColor;
	private int averageWinScoreColor;
	private int highScoreColor;

	private double lowScore;
	private double averageScore;
	private double averageWinScore;
	private double highScore;

	public ScoreGraphView(Context context) {
		super(context);
		init(context);
	}

	public ScoreGraphView(Context context, AttributeSet attrs) {
		super(context, attrs);
		init(context);
	}

	private void init(Context context) {
		barPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
		barPaint.setColor(Color.BLACK);
		barPaint.setStrokeWidth(1);

		scorePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
		scorePaint.setStrokeWidth(1);

		scoreRadius = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6, context.getResources().getDisplayMetrics());
		smallTickHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 2, context.getResources().getDisplayMetrics());
		largeTickHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6, context.getResources().getDisplayMetrics());

		lowScoreColor = ContextCompat.getColor(getContext(), R.color.score_low);
		averageScoreColor = ContextCompat.getColor(getContext(), R.color.score_average);
		averageWinScoreColor = ContextCompat.getColor(getContext(), R.color.score_average_win);
		highScoreColor = ContextCompat.getColor(getContext(), R.color.score_high);
	}

	public void setLowScore(double lowScore) {
		this.lowScore = lowScore;
	}

	public void setAverageScore(double averageScore) {
		this.averageScore = averageScore;
	}

	public void setAverageWinScore(double averageWinScore) {
		this.averageWinScore = averageWinScore;
	}

	public void setHighScore(double highScore) {
		this.highScore = highScore;
	}

	@Override
	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);

		final float y = canvas.getHeight() / 2;
		final float left = scoreRadius;
		final float right = canvas.getWidth() - scoreRadius;

		canvas.drawLine(left, y, right, y, barPaint);

		// ticks
		double tickScore = Math.ceil(lowScore / 10) * 10;
		while (tickScore <= highScore) {
			float x = (float) ((tickScore - lowScore) / (highScore - lowScore) * (right - left) + left);
			final float tickHeight = (tickScore % 50) == 0 ? largeTickHeight : smallTickHeight;
			canvas.drawLine(x, y - tickHeight, x, y + tickHeight, barPaint);
			tickScore += 10;
		}

		// score dots
		drawScore(canvas, y, left, right, lowScore, lowScoreColor);
		drawScore(canvas, y, left, right, averageScore, averageScoreColor);
		drawScore(canvas, y, left, right, averageWinScore, averageWinScoreColor);
		drawScore(canvas, y, left, right, highScore, highScoreColor);
	}

	private void drawScore(Canvas canvas, float y, float left, float right, double score, @ColorInt int color) {
		scorePaint.setColor(color);
		float x = (float) ((score - lowScore) / (highScore - lowScore) * (right - left) + left);
		canvas.drawCircle(x, y, scoreRadius, scorePaint);
	}
}

File: app/src/main/java/com/boardgamegeek/ui/widget/ScoreGraphView.kt
package com.boardgamegeek.ui.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Paint.Style
import android.util.AttributeSet
import android.util.TypedValue
import android.util.TypedValue.COMPLEX_UNIT_SP
import android.view.View
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import com.boardgamegeek.R
import com.boardgamegeek.extensions.significantDigits
import java.text.DecimalFormat
import kotlin.math.ceil

class ScoreGraphView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private val barPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val scorePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    private val scoreRadius: Float
    private val smallTickHeight: Float
    private val largeTickHeight: Float

    private val lowScoreColor: Int
    private val averageScoreColor: Int
    private val averageWinScoreColor: Int
    private val highScoreColor: Int

    var lowScore: Double = 0.toDouble()
    var averageScore: Double = 0.toDouble()
    var averageWinScore: Double = 0.toDouble()
    var highScore: Double = 0.toDouble()
    private var hasPersonalScores: Boolean = false

    var personalLowScore: Double = 0.toDouble()
        set(value) {
            field = value
            hasPersonalScores = true
        }

    var personalAverageScore: Double = 0.toDouble()
        set(value) {
            field = value
            hasPersonalScores = true
        }

    var personalAverageWinScore: Double = 0.toDouble()
        set(value) {
            field = value
            hasPersonalScores = true
        }

    var personalHighScore: Double = 0.toDouble()
        set(value) {
            field = value
            hasPersonalScores = true
        }


    init {
        barPaint.color = Color.BLACK
        barPaint.strokeWidth = 1f

        scorePaint.strokeWidth = SCORE_STROKE_WIDTH.toFloat()

        textPaint.color = ContextCompat.getColor(getContext(), R.color.secondary_text)
        textPaint.textSize = TypedValue.applyDimension(COMPLEX_UNIT_SP, 8f, getContext().resources.displayMetrics)

        scoreRadius = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6f, context.resources.displayMetrics)
        smallTickHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 2f, context.resources.displayMetrics)
        largeTickHeight = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6f, context.resources.displayMetrics)

        lowScoreColor = ContextCompat.getColor(getContext(), R.color.score_low)
        averageScoreColor = ContextCompat.getColor(getContext(), R.color.score_average)
        averageWinScoreColor = ContextCompat.getColor(getContext(), R.color.score_average_win)
        highScoreColor = ContextCompat.getColor(getContext(), R.color.score_high)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val y = (height / 2).toFloat()
        val left = scoreRadius + SCORE_STROKE_WIDTH
        val right = width.toFloat() - scoreRadius - SCORE_STROKE_WIDTH.toFloat()

        canvas.drawLine(left, y, right, y, barPaint)

        // ticks
        val scoreSpread = highScore - lowScore
        val tickSpacing = when {
            scoreSpread <= 20 -> 1
            scoreSpread <= 50 -> 5
            scoreSpread <= 200 -> 10
            scoreSpread <= 500 -> 20
            else -> (ceil(scoreSpread / 100) * 10).toInt().significantDigits(2)
        }
        var tickScore = ceil(lowScore / tickSpacing) * tickSpacing
        while (tickScore <= highScore) {
            val x = ((tickScore - lowScore) / scoreSpread * (right - left) + left).toFloat()
            val tickHeight: Float
            if (tickScore % (5 * tickSpacing) == 0.0) {
                tickHeight = largeTickHeight
                val label = SCORE_FORMAT.format(tickScore)
                val labelWidth = textPaint.measureText(label)
                val labelLeft = (x - labelWidth / 2).coerceIn(0f, width - labelWidth)
                canvas.drawText(label, labelLeft, height.toFloat(), textPaint)
            } else {
                tickHeight = smallTickHeight
            }
            canvas.drawLine(x, y - tickHeight, x, y + tickHeight, barPaint)
            tickScore += tickSpacing.toDouble()
        }

        // score dots
        scorePaint.style = if (hasPersonalScores) Style.STROKE else Style.FILL
        drawScore(canvas, y, left, right, lowScore, lowScoreColor)
        drawScore(canvas, y, left, right, highScore, highScoreColor)
        drawScore(canvas, y, left, right, averageScore, averageScoreColor)
        drawScore(canvas, y, left, right, averageWinScore, averageWinScoreColor)

        if (hasPersonalScores) {
            scorePaint.style = Style.FILL
            drawScore(canvas, y, left, right, personalLowScore, lowScoreColor)
            drawScore(canvas, y, left, right, personalHighScore, highScoreColor)
            drawScore(canvas, y, left, right, personalAverageScore, averageScoreColor)
            drawScore(canvas, y, left, right, personalAverageWinScore, averageWinScoreColor)
        }
    }

    private fun drawScore(canvas: Canvas, y: Float, left: Float, right: Float, score: Double, @ColorInt color: Int) {
        scorePaint.color = color
        val x = ((score - lowScore) / (highScore - lowScore) * (right - left) + left).toFloat()
        canvas.drawCircle(x, y, scoreRadius, scorePaint)
    }

    companion object {
        private val SCORE_FORMAT = DecimalFormat("0")
        private const val SCORE_STROKE_WIDTH = 2
    }
}
